### PR TITLE
[dynamo] treat `torch.device`, `torch.dtype` as constant literal; revise guards to have access to `torch` module

### LINKED ISSUE #109910: Dynamo error for autograd function
# Summary
Encountering Dynamo error when attempting to compile an autograd function that returns a dtype.

### Repro
``` Python
import torch

class dtype_test(torch.autograd.Function):
    @staticmethod
    def forward(
        ctx,
        tensor: torch.Tensor,
        dtype=torch.dtype,
    ):
        orig_precision = tensor.dtype
        ctx.orig_precision = orig_precision
        return tensor.to(dtype), dtype
    @staticmethod
    def backward(ctx, dOut):
        return dOut.to(ctx.orig_precision), None, None,

def main():
    x = torch.randn(16, 16, device="cpu", dtype=torch.float32)
    out = dtype_test.apply(x, torch.bfloat16)
    def test_func(x, dtype):
        return dtype_test.apply(x, dtype)
    compiled_func = torch.compile(test_func, fullgraph=True)
    y = compiled_func(x, torch.bfloat16)

if __name__ == "__main__":
    main()
 ```
 
 ### Output
 ``` Shell
  File "/home/drisspg/miniconda3/envs/nightly/lib/python3.10/site-packages/torch/_dynamo/variables/base.py", line 329, in call_method
    raise unimplemented(f"call_method {self} {name} {args} {kwargs}")
  File "/home/drisspg/miniconda3/envs/nightly/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 176, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_method TupleVariable() to [ConstantVariable(dtype)] {}

from user code:
   File "/home/drisspg/meta/float8_experimental/../scripts/compile/autograd_dtype_float.py", line 21, in test_func
    return dtype_test.apply(x, dtype)
  File "/home/drisspg/meta/float8_experimental/../scripts/compile/autograd_dtype_float.py", line 12, in forward
    return tensor.to(dtype), dtype
  ```

#### Note
This is a proxy issue I am creating for https://github.com/pytorch-labs/float8_experimental/issues/108

 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by jon-chuang**
Just like e.g. container - list/set of constant literals, these are constant literals.

We follow up to https://github.com/pytorch/pytorch/pull/112416, enforcing that we always use `ConstantVariable` to represent these.

Replace https://github.com/pytorch/pytorch/pull/112284, https://github.com/pytorch/pytorch/pull/112332 as incomplete, in case there is no movement there.

Ought to fix: https://github.com/pytorch/pytorch/issues/109910

We remove old guards special-casing, which fell back on str equality when not having access to `torch` module in `eval`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by jon-chuang** [comment]
Looks like this doesn't work (as is).

The reason is that we don't actually want to treat containers containing `dtype` and `device` as constant literals as this will make our guard logic complicated.

**Comment by jon-chuang** [comment]
Actually, there was an easy way to fix this - add torch as a closure var to the scope.

**Comment by jon-chuang** [comment]
Seems that the fsdp test failures are real. Trouble is - I can't reproduce locally, while the logs from CI are completely opaque.

**Comment by jon-chuang** [comment]
Actually, I think the fsdp test is broken. On main, it doesn't compile SDPA, this PR enables that, but when it actually compiles SDPA, it is broken.

**Review by lezcano — COMMENTED** [review]
This should be done in the previous PR

File: torch/_dynamo/variables/constant.py [hunk 1]
```diff
@@ -102,7 +102,16 @@ def getitem_const(self, arg: VariableTracker):
...
-        if type(obj) in (int, float, bool, type(None), str, Ellipsis.__class__):
+        if type(obj) in (
+            int,
+            float,
+            bool,
+            type(None),
+            str,
+            Ellipsis.__class__,
+            torch.dtype,
+            torch.device,
```
Comment (lezcano) [inline torch/_dynamo/variables/constant.py:113]: + torch.layout

Comment (jon-chuang) [reply torch/_dynamo/variables/constant.py:113]: don't really want to increase the surface area if not strictly necessary...

**Comment by jon-chuang** [comment]
> This should be done in the previous PR

Hmm, well they are related. Would have stacked them but the changes are orthogonal actually.

File: torch/_dynamo/variables/constant.py (hunk 1 above)
Comment (lezcano) [reply torch/_dynamo/variables/constant.py:113]: I mean, the three are the same, there is no reason to not handle them consistently all throughout.

File: torch/_dynamo/guards.py [hunk 2]
```diff
@@ -127,6 +127,7 @@ def uninteresting_files():
             if isinstance(a, (np.generic, np.ndarray))
             else a,
         ),
+        ("torch", torch),
```
Comment (lezcano) [inline torch/_dynamo/guards.py:130]: currently we have a global `device` set to be `torch.device`. I wonder whether just exposing torch is alright. I'll  let the others reveiw this one

File: torch/_dynamo/variables/constant.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/constant.py:113]: yes, but there are plenty of other "literals" (e.g. enums, other torch vars ) that we do not include in this codepath. So I'd prefer to just support the important ones `torch.dtype`, `torch.device`.

See also:
https://github.com/pytorch/pytorch/blob/86196bf116a67701adb0e259cfaa35a843e128ce/torch/_dynamo/guards.py#L386

**Review by ezyang — APPROVED** [review]
Thanks

**Comment by ezyang** [comment]
@jansel you should probably take a quick look, in case you had a good reason that these should not be constants originally. They are immutable, fortunately

**Pull request closed without merging**